### PR TITLE
refactor(ci): extract `setup-tauri` action and generalize Windows builds

### PR DESCRIPTION
**Pull request by ReactorScram**
This prevents duplication for different Tauri jobs like building the release packages vs testing a debug build with mock keyring.

```[tasklist]
- [ ] Fix branch protection rules for changed tests
```

### CODE REVIEW

File: .github/actions/setup-tauri/action.yml
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml:1]: 👍

File: .github/workflows/_rust.yml [hunk 1]
```diff
@@ -25,14 +25,20 @@ jobs:
         include:
           - runs-on: ubuntu-22.04
             packages: # Intentionally blank as a package catch-all linter
+            target_os: linux
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 2]
```diff
@@ -25,14 +25,20 @@ jobs:
...
           - runs-on: macos-14
             packages: -p connlib-client-apple -p snownet
+            target_os: macos
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 3]
```diff
@@ -25,14 +25,20 @@ jobs:
...
           - runs-on: windows-2022
             packages: -p connlib-client-shared -p firezone-windows-client -p snownet
+            target_os: windows
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 4]
```diff
@@ -25,14 +25,20 @@ jobs:
...
     runs-on: ${{ matrix.runs-on }}
     steps:
       - uses: actions/checkout@v4
       - uses: ./.github/actions/setup-rust
+      - uses: ./.github/actions/setup-tauri
+        with:
+          target_os: ${{ matrix.target_os }}
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
        if: ${{ runner.os == "Windows" || runner.os == "Linux" }}
```

File: .github/workflows/_rust.yml [hunk 5]
```diff
@@ -50,29 +56,57 @@ jobs:
         include:
           - runs-on: ubuntu-20.04
             packages: -p firezone-linux-client -p firezone-gateway -p connlib-client-android -p snownet
+            target_os: linux
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 6]
```diff
@@ -50,29 +56,57 @@ jobs:
...
           - runs-on: ubuntu-22.04
             packages: -p firezone-linux-client -p firezone-gateway -p connlib-client-android -p snownet
+            target_os: linux
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 7]
```diff
@@ -50,29 +56,57 @@ jobs:
...
           - runs-on: macos-12
             packages: -p connlib-client-apple -p snownet
+            target_os: macos
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 8]
```diff
@@ -50,29 +56,57 @@ jobs:
...
           - runs-on: macos-13
             packages: -p connlib-client-apple -p snownet
+            target_os: macos
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 9]
```diff
@@ -50,29 +56,57 @@ jobs:
...
           - runs-on: macos-14
             packages: -p connlib-client-apple -p snownet
+            target_os: macos
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 10]
```diff
@@ -50,29 +56,57 @@ jobs:
...
           - runs-on: windows-2019
             packages: -p firezone-windows-client -p connlib-client-shared -p snownet
+            target_os: windows
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 11]
```diff
@@ -50,29 +56,57 @@ jobs:
...
           - runs-on: windows-2022
             packages: -p firezone-windows-client -p connlib-client-shared -p snownet
+            target_os: windows
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 12]
```diff
@@ -50,29 +56,57 @@ jobs:
...
     runs-on: ${{ matrix.runs-on }}
     steps:
       - uses: actions/checkout@v4
       - uses: ./.github/actions/setup-rust
+      - uses: ./.github/actions/setup-tauri
+        with:
+          target_os: ${{ matrix.target_os }}
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
        if: ${{ runner.os == "Windows" || runner.os == "Linux" }}
```

File: .github/workflows/_rust.yml [hunk 13]
```diff
@@ -50,29 +56,57 @@ jobs:
...
+          #- runs-on: ubuntu-20.04
+          #  target: x86_64-unknown-linux-gnu
+          #  target_os: linux
+          # Broken on 22.04 <https://github.com/firezone/firezone/issues/3699>
+          #- runs-on: ubuntu-22.04
+          #  target: x86_64-unknown-linux-gnu
+          #  target_os: linux
+          - runs-on: windows-2019
+            target: x86_64-pc-windows-msvc
+            target_os: windows
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 14]
```diff
@@ -50,29 +56,57 @@ jobs:
...
+          - runs-on: windows-2022
+            target: x86_64-pc-windows-msvc
+            target_os: windows
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 15]
```diff
@@ -114,32 +140,36 @@ jobs:
...
+          # - ubuntu-20.04
+          - windows-2019
         include:
-          - package: firezone-windows-client
-            artifact: firezone-windows-client
-            name: windows-client
+          # - runs-on: ubuntu-20.04
+          #   target_os: linux
+          - runs-on: windows-2019
+            target_os: windows
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/workflows/_rust.yml [hunk 16]
```diff
@@ -86,18 +120,10 @@ jobs:
...
-        run: Invoke-WebRequest -Uri https://go.microsoft.com/fwlink/p/?LinkId=2124703 -OutFile WebView2Installer.exe
-        # This downloads about 200 MB and takes about 5 minutes on my VM
-        # So we could fault in WebView2 from the client exe without the MSI if we needed.
-        # Currently the MSI does this and it's a little janky.
-      - name: Install WebView2
-        run: Start-Process WebView2Installer.exe -ArgumentList "/install" -Wait
+          targets: ${{ matrix.target }}
+      - uses: ./.github/actions/setup-tauri
+        with:
+          target_os: ${{ matrix.target_os }}
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
        if: ${{ runner.os == "Windows" || runner.os == "Linux" }}
```

File: .github/workflows/_rust.yml [hunk 17]
```diff
@@ -114,32 +140,36 @@ jobs:
...
+      BINARY_DEST_PATH: firezone-windows-client
       CONNLIB_LOG_UPLOAD_INTERVAL_SECS: 300
     steps:
       - uses: actions/checkout@v4
       - uses: ./.github/actions/setup-node
         with:
           node-version: '20'
       - uses: ./.github/actions/setup-rust
+      - uses: ./.github/actions/setup-tauri
         with:
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
        if: ${{ runner.os == "Windows" || runner.os == "Linux" }}
```

File: .github/workflows/_rust.yml [hunk 18]
```diff
@@ -114,32 +140,36 @@ jobs:
...
-          targets: x86_64-pc-windows-msvc
+          target_os: ${{ matrix.target_os }}
```
Comment (jamilbk) [inline .github/workflows/_rust.yml]:
```suggestion
```

File: .github/actions/setup-tauri/action.yml [hunk 19]
```diff
@@ -0,0 +1,44 @@
+# From https://tauri.app/v1/guides/getting-started/prerequisites
+name: "Setup Tauri"
+description: "Sets up the dependencies for building and testing Tauri apps"
+inputs:
+  target_os:
+    description: "`windows` or `linux` like in Rust"
+    required: true
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
```

File: .github/actions/setup-tauri/action.yml [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+runs:
+  using: "composite"
+  steps:
+    - name: Apt-get update
+      if: ${{ inputs.target_os == 'linux' }}
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
      if: ${{ runner.os == 'Linux' }}
```

File: .github/actions/setup-tauri/action.yml [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+      run: sudo apt-get update
+      shell: bash
+    - name: Install Tauri build deps
+      if: ${{ inputs.target_os == 'linux' }}
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
      if: ${{ runner.os == 'Linux' }}
```

File: .github/actions/setup-tauri/action.yml [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+      run: sudo apt-get install libwebkit2gtk-4.0-dev libssl-dev libgtk-3-dev libayatana-appindicator3-dev librsvg2-dev
+      shell: bash
+    - name: Install gnome-keyring
+      if: ${{ inputs.target_os == 'linux' }}
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
      if: ${{ runner.os == 'Linux' }}
```

File: .github/actions/setup-tauri/action.yml [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+      run: sudo apt-get install gnome-keyring
+      shell: bash
+      # This is only needed if we'll launch the Tauri GUI, so it's redundant for clippy / test
+      # This is what the Tauri CI tests use
+      # <https://github.com/tauri-apps/tauri/blob/3fb414b61ad7cfce67751230826fddfb39effec5/.github/workflows/bench.yml#L74>
+    - name: Install Tauri runtime deps
+      if: ${{ inputs.target_os == 'linux' }}
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
      if: ${{ runner.os == 'Linux' }}
```

File: .github/actions/setup-tauri/action.yml [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+      run: sudo apt-get install at-spi2-core xvfb
+      shell: bash
+    - name: Download WebView2 bootstrapper
+      if: ${{ inputs.target_os == 'windows' }}
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
      if: ${{ runner.os == 'Windows' }}
```

File: .github/actions/setup-tauri/action.yml [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+      # This is the "Evergreen" bootstrapper from Microsoft
+      # <https://developer.microsoft.com/en-us/microsoft-edge/webview2/?form=MA13LH#download>
+      # Unfortunately, this makes the test non-deterministic.
+      # Controlling the version would be difficult.
+      run: Invoke-WebRequest -Uri https://go.microsoft.com/fwlink/p/?LinkId=2124703 -OutFile WebView2Installer.exe
+      shell: pwsh
+    - name: Install WebView2
+      if: ${{ inputs.target_os == 'windows' }}
```
Comment (jamilbk) [inline .github/actions/setup-tauri/action.yml]:
```suggestion
      if: ${{ runner.os == 'Windows' }}
```

**Review by jamilbk — CHANGES_REQUESTED** [review]
LGTM, just the suggestion to simplify things by using the already -available [`runner.os`](https://docs.github.com/en/actions/learn-github-actions/contexts#runner-context)

**Comment by ReactorScram** [comment]
@jamilbk Figured there's something like that, I should have asked ChatGPT lol
I'm also going to ditch the `target` for `setup-rust` whenever possible, because it already picks MSVC on Windows and GNU on Linux, which is what I want it to do.

File: .github/workflows/_rust.yml [hunk 26]
```diff
@@ -66,13 +67,24 @@ jobs:
...
+      - uses: ./.github/actions/setup-tauri
       - run: cargo test --all-features ${{ matrix.packages }}
 
   # Runs the Windows client smoke test, built in debug mode. We can't run it in release
   # mode because of a known issue: <https://github.com/firezone/firezone/blob/456e044f882c2bb314e19cc44c0d19c5ad817b7c/rust/windows-client/src-tauri/src/client.rs#L162-L164>
-  windows-smoke-test:
-    name: windows-smoke-test
-    runs-on: windows-2019
+  gui-smoke-test:
+    name: gui-smoke-test-${{ matrix.runs-on }}
```
Comment (ReactorScram) [inline .github/workflows/_rust.yml:76]: For some reason `runner.os` wasn't working in `name` keys

**Review by jamilbk — APPROVED** [review]

**Comment by ReactorScram** [comment]
@jamilbk Since the test names changed, this might have a similar problem with branch protection as we had a few weeks ago on some other PR. 

I see Github is waiting for `build-tauri-windows-client` and `windows-smoke-test`, those are now `build-gui-windows-2019`, `gui-smoke-test-windows-2019`, and `gui-smoke-test-windows-2022`

**Pull request merged**